### PR TITLE
MSD arbitrary species input

### PR DESCRIPTION
**Pull request by camelto2**
## Proposed changes

This PR generalizes the multi-slater build path. Previously, everything was hard wired to have alpha and beta electron species. The PR adds a new build path that can handle any number of fermion species to occupy the multi slater determinants. This partially addresses #2698 in that you could now use only "alpha" electrons (useful for spinors) or additional species for other applications. 

essentially, the input (if you want variable number of species) becomes
<multidetermiant fast="yes" optimize="no"  spo_0="name0" spo_1="name1" ... spo_(N-1)="nameN-1">
where the number of species is determined by the particle set. 
 
 and (in the xml example) a similarly modified det list
 
 <detlist size="5" ... nstates="8" nc0="0" ne0="5"  nc1="0" ne1="2" nc2="0" ne2="4" ...>
 
 and the CI specifcation
    
 <ci ... qchem_coeff="-0.8586" occ0="11111000" occ1="00101000" occ2="01101010"/>
          
The changes are similar for the hdf5 input style. 

This keeps all backwards compatibility, so all current unit tests/input will continue to work.

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No. 
The new input style works and preserves the previous approach so all previous calculations/formats behave as expected. However, there does need to be an additional PR to properly generalize the MultiSlaterDeterminantFast evaluation routines to handle arbitrary sum_i c_i x (an arbitrary product of determinants). See #2947 which addresses the changes necessary for MultiSlaterDeterminantFast. 
## What systems has this change been tested on?
My workstation
## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Review by ye-luo — CHANGES_REQUESTED** [review]
Hopefully my reaction is not totally a surprise.
Once  https://github.com/QMCPACK/qmcpack/issues/2735 is solved
I don't see why we need a new input tag ngroups
Also the old builder code-path should converge with the new code path, convince me if you really think it is needed to keep two versions.

**Comment by camelto2** [comment]
> Hopefully my reaction is not totally a surprise.
> 
> 1. Once  #2735 is solved
>    I don't see why we need a new input tag ngroups. I mean ngroups can be figured out by query the electron particle set.
Sure that can be done. I was essentially just doing a first pass to get some comments. 

> 2. Also the old builder code-path should converge with the new code path, convince me if you really think it is needed to keep two versions. It seems the new code is a good move but we need to kill the old one. Also need input tests like test_spo_collection_input_LCAO_xml.cpp and test_spo_collection_input_MSD_LCAO_h5.cpp.
I don't think we need to support both paths once the evaluation and everything is cleaned up. I separated everything for now just to make sure I didn't break anything that currently worked.

**Comment by ye-luo** [comment]
@camelto2 I think we don't need to wait for the fix on the evaluation side, as long as the generic input parsing can support the legacy parsing, then it is good enough for merging with some safeguard.
our input class do support multiple tags mapped to the same variable.
mparam.add(occ, "occ1") which is handled by the loop
mparam.add(occ, "occ_up") which can be added for backward compatibility.

**Comment by camelto2** [comment]
> @camelto2 I think we don't need to wait for the fix on the evaluation side, as long as the generic input parsing can support the legacy parsing, then it is good enough for merging with some safeguard.
> our input class do support multiple tags mapped to the same variable.
> mparam.add(occ, "occ1") which is handled by the loop
> mparam.add(occ, "occ_up") which can be added for backward compatibility.

Ahh, I didn't think about that. I can update that accordingly to take care of this backward compatibility issue

**Comment by ye-luo** [comment]
Also ParticleSet::groups() returns how many kinds of quantum particles are there.

**Comment by camelto2** [comment]
I removed the need for separate build paths for the MSD fast algorithm. In cases where the number of species is two, I also allow for backwards compatibility. For example, in the hdf5 for two particle species, most hdf5 files will have "CI_Alpha" "CI_Beta". So if reading CI_0, CI_1 fails, it tries CI_Alpha and CI_Beta. Same in the xml path. 

I still have overloaded readDetList and readDetListH5, which is used for the legacy all_determinants algorithm, but that can be changed  as well  using the newer builder functions.  Updating that now

**Comment by camelto2** [comment]
I added a unit test for the xml style MSD input, with an arbitrary number particle species to make sure everything in constructed correctly. I also removed the need for overloaded readDetList and readDetListH5 so that we can reduce the amount of code. Backwards compatibility is retained for both MultiSlaterDeterminant and MultiSlaterDeterminantFast.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp [hunk 1]
```diff
@@ -204,46 +204,48 @@ WaveFunctionComponent* SlaterDetBuilder::buildComponent(xmlNodePtr cur)
       if (multislaterdet_0 || multislaterdetfast_0)
         APP_ABORT("multideterminant is already instantiated.");
 
-      std::string spo_alpha_name;
-      std::string spo_beta_name;
+      int nGroups = targetPtcl.groups();
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:207]: I found it a good habit to put a const in this case when a local short name variable is used for convenience. const protects unintended modification.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp [hunk 2]
```diff
@@ -204,46 +204,48 @@ WaveFunctionComponent* SlaterDetBuilder::buildComponent(xmlNodePtr cur)
...
+      std::vector<SPOSetPtr> spos(nGroups);
+      std::vector<std::unique_ptr<SPOSet>> spo_clones(nGroups);
 
-      if (spo_beta == nullptr)
+      for (int grp = 0; grp < nGroups; grp++)
       {
-        app_error() << "In SlaterDetBuilder: SPOSet \"" << spo_beta_name
-                    << "\" is not found. Expected for MultiSlaterDeterminant.\n";
-        abort();
+        spos[grp] = sposet_builder_factory_.getSPOSet(spoNames[grp]);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:239]: The pointer return from getSPOSet should be treated as a reference. It was kept pointer for checking nullptr. Please assign the value to a temporal pointer variable. In the current way, the SPOSet actually got destroyed when you reset the smart pointer.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp [hunk 3]
```diff
@@ -204,46 +204,48 @@ WaveFunctionComponent* SlaterDetBuilder::buildComponent(xmlNodePtr cur)
...
+        if (spos[grp] == nullptr)
+        {
+          app_error() << "In SlaterDetBuilder: SPOSet \"" << spoNames[grp]
+                      << "\" is not found. Expected for MultiSlaterDeterminant.\n";
+          abort();
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:244]: Please use barrier_and_abort.

**Review by ye-luo — COMMENTED** [review]
Great simplification. Seems only minor issues left.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp (hunk 2 above)
Comment (camelto2) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:239]: I guess I'm confused as to what you mean. The smart pointer spo_clones[grp] calls reset, but the argument is spos[grp]->makeClone(). So the spo_clones[grp] should take ownership of the pointer being returned by makeClone(), not taking ownership of spos[grp] itself.  So I don't see how the original ptr returned by getSPOSet (spos[grp]) gets destroyed.

Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:239]: reset() calls delete on the previously stored pointer. This is not desired.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp (hunk 1 above)
Comment (camelto2) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp]: done

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp (hunk 3 above)
Comment (camelto2) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp]: done

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp (hunk 2 above)
Comment (camelto2) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:239]: Sure, but spo_clones[grp] is initially a nullptr, so it doesn't actually call the delete. I guess reset implies that it  *was* storing something before, so it is bad practice. I changed it to be a bit more clear hopefully.

Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:239]: I was confused by spos and spo_clones. Could you remove spos and store the return of sposet_builder_factory_.getSPOSet in spo_tmp for nullptr check locally.
```
spo_clones.emplace_back(spo_tmp->makeClone());
```
should be OK.

**Comment by ye-luo** [comment]
deterministic-unit_test_wavefunction_sposet fails now.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp (hunk 2 above)
Comment (camelto2) [reply src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp]: done

**Review by ye-luo — APPROVED** [review]
All look good to me.

**Comment by ye-luo** [comment]
Okay to test

**Comment by ye-luo** [comment]
https://cdash.qmcpack.org/CDash/viewTest.php?onlydelta&buildid=189182 failures from last night.

**Pull request merged**